Src/ILGPU.Algorithms/MatrixOperations/MaskedSparseMatrixExtensions.cs

```diff
@@ -0,0 +1,390 @@
+    /// <param name="bView">A sparse matrix B of shape NxK (will transpose).</param>
+    /// <param name="outView">
+    /// A dense output view containing the results of the multiplication.
+    /// </param>
+    public delegate void MaskedSparseMatrixMultiplier<T, TPredicate, TStride>(
+        AcceleratorStream stream,
+        TPredicate maskPredicate,
+        ArrayView2D<T, TStride> aView,
+        SparseMatrixView<T, TStride> bView,
+        ArrayView2D<T, TStride> outView)
+        where T : unmanaged
+        where TStride : struct, IStride2D
+        where TPredicate : struct, InlineList.IPredicate<Index2D>;
+
+    /// <summary>
+    /// Specialized extensions to operate on sparse matrices while taking predicates
+    /// into account to skip specific elements.
+    /// </summary>
+    public static class MaskedSparseMatrixExtensions
+    {
+        /// <summary>
+        /// The matrix multiplication kernel that runs on the accelerated device while
+        /// using thread compaction to free masked warps.
+        /// </summary>
+        /// <param name="maskPredicate">
+        /// The input masking predicate (targeting a dense matrix of shape MxK).
+        /// </param>
+        /// <param name="aView">A dense input matrix of shape MxK.</param>
+        /// <param name="bView">A sparse matrix B of shape NxK (will transpose).</param>
+        /// <param name="outView">
+        /// A dense output view containing the results of the multiplication.
+        /// </param>
+        /// <param name="processor">An instance of an actual mul-add operation.</param>
+        internal static void MaskedSparseMatrixMultiplierKernel<
+            T,
+            TPredicate,
+            TStride,
+            TProcessor>(
+            TPredicate maskPredicate,
+            ArrayView2D<T, TStride> aView,
+            SparseMatrixView<T, TStride> bView,
+            ArrayView2D<T, TStride> outView,
+            TProcessor processor)
+            where T : unmanaged
+            where TStride : struct, IStride2D
+            where TPredicate : struct, InlineList.IPredicate<Index2D>
+            where TProcessor : struct, IMaskedSparseMatrixProcessor<T>
+        {
+            var maskView = SharedMemory.GetDynamic<int>();
+            Trace.Assert(
+                Group.DimX <= maskView.IntLength,
+                "Invalid shared memory config");
+
+            // Get all predicate results and perform thread compaction
+            int index = Grid.GlobalLinearIndex;
+            var maskIndex2D = aView.Stride.ReconstructFromElementIndex(index);
+            bool apply = maskIndex2D.X < aView.Extent.X & maskIndex2D.Y < aView.Extent.Y;
+
+            // Check the predicate (if in range) and compute thread-compaction offsets
+            if (apply)
+                apply = maskPredicate.Apply(maskIndex2D);
+            int groupOffset = GroupExtensions.ExclusiveScan<int, AddInt32>(
+                Utilities.Select(apply, 1, 0));
```

Comment:
So, if I understand this correctly you are making some assumptions about the sparsity structure defined by your predicate. Basically, you are assuming the mask will have a number of contiguous "chunks" so that you don't get too many groups here.  Otherwise, if you have a mask pattern like 0, 1, 0, 1, 0, 1, ...  I imagine this would quickly degrade.
I would suggest that we 
1. Get information from the quants about what their mask pattern looks like.
2. Document the kind of mask structure / assumption that this code is optimized for.
3. When we add unit tests show both "good" and "adversarial" cases where the performance degrades.